book-cw/part-2-coding/regex.adoc

```diff
@@ -1,3 +1,348 @@
+
+You'll note that all of these are more complicated than the non-regex search term:
+
+* `codewriting.org`
+
+Probably in most cases of a find-and-replace task, I don't get to use regex.
+
+But what if we needed to find all instances of `http://` protocol indicator for a `codewriting.org` URL in our docs and replace `http:` with `https:`.
+We don't want this for every instance of `http:` in the document, but we also cannot count on the domain string to always appear intact, as `http://codewriting.org`, which we could easily replace with `https://codewriting.org`.
+Making matters more complicated, numerous instances of a `codewriting.org` URL appear with a hostname or “subdomain”, such as `docs.codewriting.org`.
+To make sure we caught them all manually, we'd first have to create a list, and then iterate through it one by one replacing the whole string.
+
+Here's the simple regex pattern that will save us all this trouble:
+
+[source,regex]
+----
+http\:\/\/([a-z]+\.)?codewriting\.org
+----
+
+This regex would efficiently match the following strings:
+
+* `http://codewriting.org`
+* `http://git.codewriting.org`
+* `http://staging.codewriting.org`
+
+The single set of parentheses in the regex will capture any “group” it matches and store it as a variable we can later reference to reinsert the captured string.
+The `$1` in the following replace string represents the first (and only) such group captured in each of our examples.
+
+----
+https://$1codewriting.org
+----
+
+The `$1` will insert whatever was captured for each instance: either an empty string, `git.`, or `staging.`.
+Therefore, one find/replace procedure will handle the whole fix.
+
+The above pattern assumes I know that all subdomains are lowercase letters (`[a-z]+`), or it would get more complicated (`[a-zA-Z0-9][a-zA-Z0-9\-]+`).
+This pattern indicates a single alphanumeric character followed by any combination of alphanumeric or the `-` symbol, the only allowable format for hostnames.
+
+Let's say for some reason we also have various top-level domains in addition to different subdomains.
+Maybe we also bought `codewriting.net` and have different apps running on different hosts, and our docs need to reflect the distinction.
+
+We need to upgrade our pattern to catch these new variations.
+
+[source,regex]
+----
+http\:\/\/([a-zA-Z0-9][a-zA-Z0-9\-]+\.)?codewriting\.([a-zA-Z]{2,})
+----
+
+Now we can match several new patterns that we were unable to previously, along with a second group for reflecting the top-level domain.
+
+* `http://codewriting.org`
+* `http://git.codewriting.org`
+* `http://staging.codewriting.org`
+* `http://shop.codewriting.com`
+* `http://members.codewriting.net`
+* `http://dating.codewriting.net`
+* `http://codewriting.ca`
+* `http://1staging.codewriting.org`
+
+Our replace string is still pretty simple, with the second capture group token added in place of the definite `org`:
+
+----
+https://$1codewriting.$2
+----
+
+== Why (and When) to Use Regex
```

Comment:
Recommend that you move the "Why" of regex up higher, or else the first section is mostly "regex is hard." You might also separate the "Why regex is useful" info from the cost-benefit analysis of automation, because they seem like different topics to me.